File: ios/Sources/GutenbergKit/Sources/RESTAPIRepository.swift
import CryptoKit
import Foundation

/// A caching repository for WordPress REST API resources needed by the editor.
///
/// `RESTAPIRepository` handles fetching and caching API responses such as editor settings,
/// post data, theme information, and site options. Cached responses are stored on disk
/// and returned on subsequent requests to improve loading performance.
public struct RESTAPIRepository: Sendable {

    let httpClient: EditorHTTPClientProtocol
    private let configuration: EditorConfiguration
    private let cache: EditorURLCache

    // Other Endpoints
    private let editorSettingsUrl: URL
    private let activeThemeUrl: URL
    private let siteSettingsUrl: URL
    private let postTypesUrl: URL

    public init(
        configuration: EditorConfiguration,
        httpClient: EditorHTTPClientProtocol,
        urlSession: URLSession = .shared,
        cache: EditorURLCache
    ) {
        self.httpClient = httpClient
        self.configuration = configuration
        self.cache = cache

        let apiRoot = configuration.siteApiRoot

        // Use custom endpoint if provided, otherwise build from apiRoot with namespace
        if let customEndpoint = configuration.editorSettingsEndpoint {
            self.editorSettingsUrl = customEndpoint
        } else {
            self.editorSettingsUrl = Self.buildNamespacedURL(
                apiRoot: apiRoot,
                path: Constants.API.editorSettingsPath,
                namespace: configuration.siteApiNamespace.first
            )
        }

        self.activeThemeUrl = Self.buildNamespacedURL(
            apiRoot: apiRoot,
            path: Constants.API.activeThemePath,
            namespace: configuration.siteApiNamespace.first
        )
        self.siteSettingsUrl = Self.buildNamespacedURL(
            apiRoot: apiRoot,
            path: Constants.API.siteSettingsPath,
            namespace: configuration.siteApiNamespace.first
        )
        self.postTypesUrl = Self.buildNamespacedURL(
            apiRoot: apiRoot,
            path: Constants.API.postTypesPath,
            namespace: configuration.siteApiNamespace.first
        )
    }

    /// Builds a URL by inserting the namespace after the version segment of the path.
    /// For example: `/wp/v2/posts` with namespace `sites/123/` becomes `/wp/v2/sites/123/posts`
    private static func buildNamespacedURL(apiRoot: URL, path: String, namespace: String?) -> URL {
        guard let rawNamespace = namespace else {
            return apiRoot.appending(rawPath: path)
        }

        let namespace = rawNamespace.hasSuffix("/") ? rawNamespace : rawNamespace + "/"

        // Parse the path to find where to insert the namespace
        // Path format is typically: /prefix/version/endpoint (e.g., /wp/v2/posts or /wp-block-editor/v1/settings)
        let components = path.split(separator: "/", omittingEmptySubsequences: true)
        guard components.count >= 2 else {
            return apiRoot.appending(rawPath: path)
        }

        // Insert namespace after the version segment (second component)
        // e.g., /wp-block-editor/v1/settings -> /wp-block-editor/v1/sites/123/settings
        let prefix = components[0]
        let version = components[1]
        let remainder = components.dropFirst(2).joined(separator: "/")

        let namespacedPath: String
        if remainder.isEmpty {
            namespacedPath = "/\(prefix)/\(version)/\(namespace)"
        } else {
            namespacedPath = "/\(prefix)/\(version)/\(namespace)\(remainder)"
        }

        return apiRoot.appending(rawPath: namespacedPath)
    }

    /// Clears all cached API responses.
    public func purge() throws {
        try self.cache.clear()
    }

    // MARK: Post
    @discardableResult
    public func fetchPost(id: Int) async throws -> EditorURLResponse {
        let request = URLRequest(method: .GET, url: self.buildPostUrl(id: id))
        let response = try await self.httpClient.perform(request)
        return EditorURLResponse(response)
    }

    public func readPost(id: Int) throws -> EditorURLResponse? {
        try self.cache.response(for: buildPostUrl(id: id), httpMethod: .GET)
    }

    private func buildPostUrl(id: Int) -> URL {
        let restNamespace = configuration.postType.restNamespace
        let restBase = configuration.postType.restBase
        return Self.buildNamespacedURL(
            apiRoot: configuration.siteApiRoot,
            path: "/\(restNamespace)/\(restBase)/\(id)",
            namespace: configuration.siteApiNamespace.first
        ).appending(queryItems: [
            URLQueryItem(name: "context", value: "edit")
        ])
    }

    // MARK: Editor Settings
    @discardableResult
    public func fetchEditorSettings() async throws -> EditorSettings {
        if !self.configuration.shouldUseThemeStyles {
            return .undefined
        }

        let request = URLRequest(method: .GET, url: editorSettingsUrl)
        let response = try await self.httpClient.perform(request)

        let editorSettings = try EditorSettings(data: response.0)

        let urlResponse = EditorURLResponse((try JSONEncoder().encode(editorSettings), response.1))
        try self.cache.store(urlResponse, for: editorSettingsUrl, httpMethod: .GET)
        return editorSettings
    }

    public func readEditorSettings() throws -> EditorSettings? {
        guard let editorURLResponse = try self.cache.response(for: editorSettingsUrl, httpMethod: .GET) else {
            return nil
        }

        return try JSONDecoder().decode(EditorSettings.self, from: editorURLResponse.data)
    }

    // MARK: GET Post Type
    @discardableResult
    func fetchPostType(for type: String) async throws -> EditorURLResponse {
        try await self.perform(method: .GET, url: self.buildPostTypeUrl(type: type))
    }

    func readPostType(for type: String) throws -> EditorURLResponse? {
        try self.cache.response(for: buildPostTypeUrl(type: type), httpMethod: .GET)
    }

    private func buildPostTypeUrl(type: String) -> URL {
        Self.buildNamespacedURL(
            apiRoot: configuration.siteApiRoot,
            path: "/wp/v2/types/\(type)",
            namespace: configuration.siteApiNamespace.first
        ).appending(queryItems: [
            URLQueryItem(name: "context", value: "edit")
        ])
    }

    // MARK: GET Active Theme
    @discardableResult
    func fetchActiveTheme() async throws -> EditorURLResponse {
        try await self.perform(method: .GET, url: self.activeThemeUrl)
    }

    func readActiveTheme() throws -> EditorURLResponse? {
        try self.cache.response(for: self.activeThemeUrl, httpMethod: .GET)
    }

    // MARK: OPTIONS Settings
    @discardableResult
    func fetchSettingsOptions() async throws -> EditorURLResponse {
        try await self.perform(method: .OPTIONS, url: self.siteSettingsUrl)
    }

    func readSettingsOptions() throws -> EditorURLResponse? {
        try self.cache.response(for: self.siteSettingsUrl, httpMethod: .OPTIONS)
    }

    // MARK: Post Types
    @discardableResult
    func fetchPostTypes() async throws -> EditorURLResponse {
        try await self.perform(method: .GET, url: self.postTypesUrl)
    }

    func readPostTypes() throws -> EditorURLResponse? {
        try self.cache.response(for: self.postTypesUrl, httpMethod: .GET)
    }

    private func perform(method: EditorHttpMethod, url: URL) async throws -> EditorURLResponse {
        let request = URLRequest(method: method, url: url)
        let response = try await self.httpClient.perform(request)
        let urlResponse = EditorURLResponse(response)
        try self.cache.store(urlResponse, for: url, httpMethod: method)
        return urlResponse
    }

}

File: android/Gutenberg/src/main/java/org/wordpress/gutenberg/RESTAPIRepository.kt
package org.wordpress.gutenberg

import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.wordpress.gutenberg.model.EditorConfiguration
import org.wordpress.gutenberg.model.EditorSettings
import org.wordpress.gutenberg.model.http.EditorHTTPHeaders
import org.wordpress.gutenberg.model.http.EditorHttpMethod
import org.wordpress.gutenberg.model.http.EditorURLResponse
import org.wordpress.gutenberg.stores.EditorURLCache

/**
 * A caching repository for WordPress REST API resources needed by the editor.
 *
 * `RESTAPIRepository` handles fetching and caching API responses such as editor settings,
 * post data, theme information, and site options. Cached responses are stored on disk
 * and returned on subsequent requests to improve loading performance.
 */
class RESTAPIRepository(
    private val configuration: EditorConfiguration,
    val httpClient: EditorHTTPClientProtocol,
    private val cache: EditorURLCache
) {
    private val json = Json { ignoreUnknownKeys = true }

    private val editorSettingsUrl = buildNamespacedUrl(EDITOR_SETTINGS_PATH)
    private val activeThemeUrl = buildNamespacedUrl(ACTIVE_THEME_PATH)
    private val siteSettingsUrl = buildNamespacedUrl(SITE_SETTINGS_PATH)
    private val postTypesUrl = buildNamespacedUrl(POST_TYPES_PATH)

    /**
     * Cleanup any expired cache entries.
     *
     */
    fun cleanup() {
        cache.clean()
    }

    /**
     * Clears all cached API responses.
     */
    fun purge() {
        cache.purge()
    }

    // MARK: Post

    /**
     * Fetches post data for the given post ID.
     *
     * @param id The post ID to fetch.
     * @return The response containing the post data.
     */
    suspend fun fetchPost(id: Int): EditorURLResponse {
        val url = buildPostUrl(id)
        val response = httpClient.perform(EditorHttpMethod.GET, url)
        return EditorURLResponse(
            data = response.stringData,
            responseHeaders = response.headers
        )
    }

    /**
     * Reads cached post data for the given post ID.
     *
     * @param id The post ID to look up.
     * @return The cached response, or `null` if not cached.
     */
    fun readPost(id: Int): EditorURLResponse? {
        return cache.getResponse(buildPostUrl(id), EditorHttpMethod.GET)
    }

    private fun buildPostUrl(id: Int): String {
        val restNamespace = configuration.postType.restNamespace
        val restBase = configuration.postType.restBase
        return buildNamespacedUrl("/$restNamespace/$restBase/$id?context=edit")
    }

    // MARK: Editor Settings

    /**
     * Fetches editor settings from the WordPress REST API.
     *
     * Returns [EditorSettings.undefined] if plugins and theme styles are both disabled
     * in the configuration.
     *
     * @return The parsed editor settings.
     */
    suspend fun fetchEditorSettings(): EditorSettings {
        if (!configuration.themeStyles) {
            return EditorSettings.undefined
        }

        val response = httpClient.perform(EditorHttpMethod.GET, editorSettingsUrl)
        val editorSettings = EditorSettings.fromData(response.stringData)

        // Store the parsed settings in cache
        val urlResponse = EditorURLResponse(
            data = json.encodeToString(editorSettings),
            responseHeaders = response.headers
        )
        cache.store(urlResponse, editorSettingsUrl, EditorHttpMethod.GET)

        return editorSettings
    }

    /**
     * Reads cached editor settings.
     *
     * @return The cached settings, or `null` if not cached.
     */
    fun readEditorSettings(): EditorSettings? {
        val response = cache.getResponse(editorSettingsUrl, EditorHttpMethod.GET) ?: return null

        return runCatching {
            json.decodeFromString<EditorSettings>(response.data)
        }.getOrNull()
    }

    // MARK: GET Post Type

    /**
     * Fetches the schema for a specific post type.
     *
     * @param type The post type slug (e.g., "post", "page").
     * @return The response containing the post type schema.
     */
    internal suspend fun fetchPostType(type: String): EditorURLResponse {
        return perform(EditorHttpMethod.GET, buildPostTypeUrl(type))
    }

    /**
     * Reads cached post type schema.
     *
     * @param type The post type slug to look up.
     * @return The cached response, or `null` if not cached.
     */
    internal fun readPostType(type: String): EditorURLResponse? {
        return cache.getResponse(buildPostTypeUrl(type), EditorHttpMethod.GET)
    }

    private fun buildPostTypeUrl(type: String): String {
        return buildNamespacedUrl("/wp/v2/types/$type?context=edit")
    }

    // MARK: GET Active Theme

    /**
     * Fetches the active theme information.
     *
     * @return The response containing the active theme data.
     */
    internal suspend fun fetchActiveTheme(): EditorURLResponse {
        return perform(EditorHttpMethod.GET, activeThemeUrl)
    }

    /**
     * Reads cached active theme information.
     *
     * @return The cached response, or `null` if not cached.
     */
    internal fun readActiveTheme(): EditorURLResponse? {
        return cache.getResponse(activeThemeUrl, EditorHttpMethod.GET)
    }

    // MARK: OPTIONS Settings

    /**
     * Fetches site settings options (OPTIONS request).
     *
     * @return The response containing the settings schema.
     */
    internal suspend fun fetchSettingsOptions(): EditorURLResponse {
        return perform(EditorHttpMethod.OPTIONS, siteSettingsUrl)
    }

    /**
     * Reads cached settings options.
     *
     * @return The cached response, or `null` if not cached.
     */
    internal fun readSettingsOptions(): EditorURLResponse? {
        return cache.getResponse(siteSettingsUrl, EditorHttpMethod.OPTIONS)
    }

    // MARK: Post Types

    /**
     * Fetches all available post types.
     *
     * @return The response containing the post types data.
     */
    internal suspend fun fetchPostTypes(): EditorURLResponse {
        return perform(EditorHttpMethod.GET, postTypesUrl)
    }

    /**
     * Reads cached post types.
     *
     * @return The cached response, or `null` if not cached.
     */
    internal fun readPostTypes(): EditorURLResponse? {
        return cache.getResponse(postTypesUrl, EditorHttpMethod.GET)
    }

    private suspend fun perform(method: EditorHttpMethod, url: String): EditorURLResponse {
        val response = httpClient.perform(method, url)
        val urlResponse = EditorURLResponse(
            data = response.stringData,
            responseHeaders = response.headers
        )
        cache.store(urlResponse, url, method)
        return urlResponse
    }

    /** Builds a namespaced REST URL via the shared [RestUrlBuilder]. */
    private fun buildNamespacedUrl(path: String): String =
        RestUrlBuilder.namespaced(
            configuration.siteApiRoot,
            configuration.siteApiNamespace.firstOrNull(),
            path
        )

    companion object {
        private const val EDITOR_SETTINGS_PATH = "/wp-block-editor/v1/settings"
        private const val ACTIVE_THEME_PATH = "/wp/v2/themes?context=edit&status=active"
        private const val SITE_SETTINGS_PATH = "/wp/v2/settings"
        private const val POST_TYPES_PATH = "/wp/v2/types?context=view"
    }
}
